core/src/com/google/inject/internal/CycleDetectingLock.java

```diff
@@ -249,6 +249,9 @@ private ListMultimap<Thread, ID> detectPotentialLocksCycle() {
             MultimapBuilder.linkedHashKeys().arrayListValues().build();
         // lock that is a part of a potential locks cycle, starts with current lock
         ReentrantCycleDetectingLock<?> lockOwnerWaitingOn = this;
+        // Its possible the lock owner thread has not removed itself yet from the waiting-on-lock list.
+        // So we remove it here, to prevent an endless loop. See #1510.
+        lockThreadIsWaitingOn.remove(lockOwnerWaitingOn.lockOwnerThread);
```

Comment:
`this.remove(this.lockOwnerThread);`  ?

Reply:
You mean ```lockThreadIsWaitingOn.remove(this.lockOwnerThread)```?

Sure, why not. I prefer to use ```lockOwnerWaitingOn``` after this is stored into it, but its whichever.